Add CRS option and refactor write_datasource

ondave opened the pull request:
Add an option to define a CRS for a datasource which may not have the crs internally defined. For a new datasource on an overwrite this CRS transformation is also applied to the datasource geometry.

rafa-guedes commented:
I have tested this with a geotiff in NZGD2000 and it worked nicely. It set crs in the schema and the appropriate geometry so it was much easier to ingest this datasource.

rafa-guedes commented:
Couple of notes:

I first tried to ingest this without setting crs in the hope it would pick it up from the spatial_ref coord in the DataArray read from rioxarray.open_rasterio. However I was not able to ingest this because DataArrays are not supported:

`DatameshWriteError: 'DataArray' object has no attribute 'to_parquet'`

I then converted this into a Dataset and ingested it without setting crs. spatial_ref now became a coordinate in the dataset:

```
<xarray.Dataset> Size: 59kB
Dimensions:      (band: 1, x: 99, y: 144)
Coordinates:
  * band         (band) int64 8B 1
  * x            (x) float64 792B 1.1e+06 1.11e+06 ... 2.07e+06 2.08e+06
  * y            (y) float64 1kB 6.183e+06 6.173e+06 ... 4.763e+06 4.753e+06
    spatial_ref  int64 8B 0
Data variables:
    dummy        (band, y, x) float32 57kB ...
```

This data was ingested but the crs was not handled. I actually ended up breaking the Datamesh UI when trying to view it because latitude was not between -90 and 90

Finally, setting crs worked great. The two questions I'd have from this are:

- Should we support ingesting a DataArray? GeoTiff files are read as DataArray using rasterio so it would probably make it easier.
- How can we specify the crs in the xarray object so it is understood by Datamesh when ingesting it so we don't need to explicitly provide it?

ondave commented:
Added extra check to make sure we are dealing with a Dataset not a DataArray. It is easy to wrap those as you did. I don't think we want to support DataArrays directly at the moment without more testing what could happen in various cases.

Also now added a check for a CRS in a new datasource which will become default CRS for the subsequent geometry. This can still be overridden with the actual crs argument.
